cocotb/types/array.py

```diff
@@ -1,15 +1,116 @@
 # Copyright cocotb contributors
 # Licensed under the Revised BSD License, see LICENSE for details.
 # SPDX-License-Identifier: BSD-3-Clause
-from typing import Optional, Any, Iterable, Iterator, overload
-from collections.abc import Sequence
-from .range import Range
-from sys import maxsize
+import typing
+from abc import ABC, abstractmethod, abstractproperty
+from itertools import chain
 
+from cocotb.types.range import Range
 
-class Array(Sequence):
+T = typing.TypeVar("T")
+S = typing.TypeVar("S")
+
+
+class ArrayABC(typing.Reversible[T], typing.Collection[T], ABC):
+    """ """
+
+    @abstractproperty
+    def range(self) -> Range:
+        ...
+
+    @range.setter
+    def range(self, range: Range) -> None:
+        ...
+
+    @typing.overload
+    def __getitem__(self, item: int) -> T:
+        ...
+
+    @typing.overload
+    def __getitem__(self, item: slice) -> "ArrayABC[T]":
+        ...
+
+    @abstractmethod
+    def __getitem__(self, item):  # type: ignore
+        ...
+
+    @typing.overload
+    def __setitem__(self, item: int, value: T) -> None:
+        ...
+
+    @typing.overload
+    def __setitem__(self, item: slice, value: typing.Iterable[T]) -> None:
+        ...
+
+    @abstractmethod
+    def __setitem__(self, item, value):  # type: ignore
+        ...
```

Comment:
I think since this is an actual method that can be called and not just a stub for typing, this should probably `raise NotImplementedError`

Reply:
Will probably add `raise NotImplementedError` to the list of exclude_lines in #2624.